### ISSUE TITLE
Detect which method is available in php version?

### ISSUE THREAD

**Issue by roukmoute** [vimeo/psalm#2809]
With this example:

```php
class DatetimeImmutable extends \DateTimeImmutable
{
    public static function createFromInterface(\DatetimeInterface $datetime): \DatetimeImmutable
    {
        if (\PHP_VERSION_ID >= 80000) {
            return parent::createFromInterface($datetime);
        }
        //…
    }
}
```

psalm returns to me:
```
ERROR: UndefinedMethod - src/Datetime.php - Method DateTime::createfrominterface does not exist
            return parent::createFromInterface($datetime);
```

Is it a bug? Or I must ignore it because psalm can't detect this kind ot things.

**Comment by roukmoute** [comment]
Here it is a link: https://psalm.dev/r/c6f2bbb578

**Comment by pilif** [comment]
somebody (maybe you?) will have to add declarations for methods added to php 8 in a file similar to [CallMap_74_delta.php](https://github.com/vimeo/psalm/blob/dc78e9c57ba58e21e7970451ca52314b398fc811/src/Psalm/Internal/CallMap_74_delta.php), but named `CallMap_80_delta.php`

**Comment by orklah** [comment]
@weirdan I wonder if this was fixed by your recent PR about conditional execution?

**Comment by weirdan** [comment]
@orklah I decided to play it safe and only enabled version-conditional logic during autoload registration. This could probably be allowed to work during other execution phases too, but needs testing.

**Comment by dkarlovi** [comment]
@orklah @weirdan we discussed this yesterday in Symfony Slack, I guess no need to open another ticket. :)

**Comment by weirdan** [comment]
@dkarlovi as a smoke test, would you mind making that code piece I linked to on Slack unconditional and running analysis on your project?

**Comment by dkarlovi** [comment]
@weirdan can you point me to a place where you do it only for files which are in Composer autoload list or something? I've tried dumping `$file_path` on top of the method and the only of my files it reaches is `src/bootstrap.php` (which is only referenced from PHPUnit's config). The other files are similar, 

```
/project/vendor/phpunit/phpunit/src/Framework/Assert/Functions.php
/project/vendor/symfony/deprecation-contracts/function.php
/project/vendor/symfony/polyfill-php80/bootstrap.php
```

etc.

**Comment by weirdan** [comment]
I mean drop the `if()` and make it unconditional here:

https://github.com/vimeo/psalm/blob/64df00baba61af63af4cf42ab1489cc43a959974/src/Psalm/Internal/PhpVisitor/Reflector/ExpressionResolver.php#L307-L338

`register_autoload_files` flag is set when Psalm scans the files included during autoload registration.

**Comment by dkarlovi** [comment]
Right, my point is this method never gets invoked for the file which contains the version guarded block. I can remove the first if, it does nothing because it's never triggered. It's just triggered for a bunch of files in vendor and my bootstrap.php which isn't relevant here.
